crates/nu-protocol/src/value/lazy_record.rs

```diff
@@ -0,0 +1,57 @@
+use crate::{ShellError, Span, Value};
+use std::{cmp::Ordering, collections::HashMap, fmt};
+
+// Trait definition for a lazy record (where columns are evaluated on-demand)
+// typetag is needed to make this implement Serialize+Deserialize... even though we should never actually serialize a LazyRecord.
+// To serialize a LazyRecord, collect it into a Value::Record first.
+#[typetag::serde(tag = "type")]
+pub trait LazyRecord: fmt::Debug + Send + Sync {
+    // Get a map of
+    fn get_column_map(&self) -> HashMap<String, Box<dyn Fn() -> Result<Value, ShellError> + '_>>;
+
+    // Convert the lazy record into a regular Value::Record by collecting all its columns
+    // This is used to represent the custom value using the table representations
+    // That already exist in nushell
+    fn collect(&self) -> Result<Value, ShellError> {
```

Comment:
Hi, I've a little thought about the signature of `collect` function, maybe it's better to consume `self`?  

So we can avoid extra clone inside `map.iter`, and once we convert to `Record`, we may no-longer need to keep the original LazyRecord.

Reply:
That's a good idea. I'm not sure how to do it (maybe with a `Cell`?).

Reply:
Nope, what I mean is just like the following:
```rust
// use `self` instead of `&self` to consume self
fn collect(self) -> Result<Value, ShellError> {
        let map = self.get_column_map();

        let mut cols = vec![];
        let mut vals = vec![];

        for (column_name, closure) in map.into_iter() {
            cols.push(column_name);   // no need to clone
            let value = closure()?;
            vals.push(value);
        }

        Ok(Value::Record {
            cols,
            vals,
            span: self.span(),
        })
    }
```

Not tested by myself, but I think it's ok..

Reply:
Unfortunately I don't think that's going to work. Compilation fails because we don't know the size of every `LazyRecord` implementation:

```
error[E0277]: the size for values of type `Self` cannot be known at compilation time
  --> crates/nu-protocol/src/value/lazy_record.rs:18:16
   |
18 |     fn collect(self) -> Result<Value, ShellError> {
   |                ^^^^ doesn't have a size known at compile-time
```

Reply:
Feel free to reopen this if you think there's a good way to handle this.